src/cdomains/lval.ml

```diff
@@ -112,8 +112,7 @@ struct
   let rec leq x y =
     match x, y with
     | `NoOffset, `NoOffset -> true
-    | `NoOffset, x -> cmp_zero_offset x <> `MustNonzero
-    | x, `NoOffset -> cmp_zero_offset x = `MustZero
+    | `NoOffset, x -> true
```

Comment:
This change was made in ff03d1c22661da516c453317de35e3d44076bceb in response to https://github.com/goblint/analyzer/issues/101#issuecomment-1200996946., but this might not be entirely correct indeed.
Since I changed `merge` to be independent of `cmp_zero_offset`, it also seems like `leq` should be. Maybe a more delicate fix is needed though.

Reply:
I agree that this isn't right, because `a` (which is equivalent to `a[0]`) shouldn't be leq of `a[1]`. The modified `join` behavior is then wrong too, because `a` join `a[1]` shouldn't be `a[1]`, but `a[0..1]` or something.

Meanwhile `a` is also equivalent to `a.fst` (first field), which is fully incomparable with `a.snd` (second field). So I'm now completely unsure of what the lattice and partitioning of individual addresses (and their offsets) is (supposed to be).

Especially because `a.snd` at runtime actually corresponds to, e.g. `a[4]` based on the field widths and paddings. So should they be also considered equivalent in the lattice (and thus belong to the same partition)? If so, then all field offsets are also joinable: `a.fst` join `a.snd` is equivalent to `a[0]` join `a[4]` which could be `a[0..4]`. But right now we consider distinct fields totally separate.

This behavior would be inevitable if we simplify the partitions to be per `varinfo` (ignoring all the offsets), since the force-joined set of `a.fst` and `a.snd`, which currently is `{a.fst, a.snd}` would then become an useless `{a[0..4]}`.